### ISSUE TITLE
josephdara - No access control with public visibility for  ``` function createDToken```

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-06-dodo-judging#181]
josephdara

medium

# No access control with public visibility for  ``` function createDToken```

## Summary
In the ```D3Vault.sol``` there are 3 functions used for creation and setting of tokens in the vault, 
- ``` function addNewToken```
- ```  function createDToken```
- ```function setToken```
The first function ``` addNewToken``` is an external function with the onlyOwner modifier used to create a d3token and add it to the tokens array. It calls ```createDToken``` to create a new D3Tokeninstance
The second function ```createDToken``` is a public function with no access control modifier, it is used to generate a new instance of a D3Token and initialize it but does not add it to the list.
The third function ``` setToken``` is used to modify info of contracts already on the list of tokens.

Going by the design pattern, ```  function createDToken``` is supposed to be private or it should have an onlyOwner modifier.
It should be private because it is only called once - inside ```addNewToken``` throughout the codebase. And also tokens created with ```createDToken``` can  never be added to the vault because ```setToken``` requires the token to be added already.
## Vulnerability Detail
private function made public with no modifier 
## Impact
https://github.com/sherlock-audit/2023-06-dodo/blob/main/new-dodo-v3/contracts/DODOV3MM/D3Vault/D3Vault.sol#L153-L158
https://github.com/sherlock-audit/2023-06-dodo/blob/main/new-dodo-v3/contracts/DODOV3MM/D3Vault/D3Vault.sol#L128-L136
https://github.com/sherlock-audit/2023-06-dodo/blob/main/new-dodo-v3/contracts/DODOV3MM/D3Vault/D3Vault.sol#L159-L167
## Tool used

Manual Review

## Recommendation
Make the function private
```solidity
    function createDToken(address token) private returns (address) {
        address d3Token = ICloneFactory(_CLONE_FACTORY_).clone(_D3TOKEN_LOGIC_);
        IDToken(d3Token).init(token, address(this));
        return d3Token;
    }
```

**Comment by Josephdara** [comment]
Escalate
My reason provided did not include withdrawing with fake tokens
> dToken created by user cannot be used to withdraw underlying token.

The bug stated here was leaving the ```function createDToken```, it is obvious that it cannot be used anywhere in the protocol, they cannot be minted in the first place. 
However, the function being public was a bug as it is not called anywhere except in ``` function addNewToken```. And tokens created by the vault owner through ``` function createDToken(address token)``` can never be added to the vault, as stated in my explanation.
> It should be private because it is only called once - inside addNewToken throughout the codebase. And also tokens created with createDToken can never be added to the vault because setToken requires the token to be added already.

A token not in the ```mapping(address => bool) public tokens;``` cannot be set.
Therefore I saw that this function was redundant and was actually supposed to be private

**Comment by sherlock-admin2** [comment]
> Escalate
> My reason provided did not include withdrawing with fake tokens
> > dToken created by user cannot be used to withdraw underlying token.
> 
> The bug stated here was leaving the ```function createDToken```, it is obvious that it cannot be used anywhere in the protocol, they cannot be minted in the first place. 
> However, the function being public was a bug as it is not called anywhere except in ``` function addNewToken```. And tokens created by the vault owner through ``` function createDToken(address token)``` can never be added to the vault, as stated in my explanation.
> > It should be private because it is only called once - inside addNewToken throughout the codebase. And also tokens created with createDToken can never be added to the vault because setToken requires the token to be added already.
> 
> A token not in the ```mapping(address => bool) public tokens;``` cannot be set.
> Therefore I saw that this function was redundant and was actually supposed to be private
> 

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by hrishibhat** [comment]
Result:
Low
Has duplicates
Do not see an issue here. Having a public modifier instead of private which does not affect the protocol in anyway is not a valid high/medium
https://docs.sherlock.xyz/audits/judging/judging

**Comment by sherlock-admin2** [comment]
Escalations have been resolved successfully!

Escalation status:
- [josephdara](https://github.com/sherlock-audit/2023-06-dodo-judging/issues/181/#issuecomment-1648296168): rejected

**Comment by IAm0x52** [comment]
Fixed in [PR#50](https://github.com/DODOEX/new-dodo-v3/pull/50)

Fix looks good. createDToken is now internal instead of being public
